fw/http_parser.c

```diff
@@ -2312,65 +2420,102 @@ __req_parse_cache_control(TfwHttpReq *req, unsigned char *data, size_t len)
 	__FSM_REQUIRE_FIRST_DIGIT(Req_I_CC_MinFreshVBeg, Req_I_CC_MinFreshV);
 
 	__FSM_STATE(Req_I_CC_MinFreshV) {
+		parser->cache_control.filled = true;
+		parser->cache_control.dir_allowed = false;
+
 		__fsm_sz = __data_remain(p);
 		__fsm_n = parse_uint_list(p, __fsm_sz, &parser->_acc);
 		if (__fsm_n == CSTR_POSTPONE)
 			__msg_hdr_chunk_fixup(data, len);
 		if (__fsm_n < 0)
-			return __fsm_n;
+			__FSM_EXIT(__fsm_n);
 		req->cache_ctl.min_fresh = parser->_acc;
 		req->cache_ctl.flags |= TFW_HTTP_CC_MIN_FRESH;
 		__FSM_I_MOVE_n(Req_I_EoT, __fsm_n);
 	}
 
 	__FSM_STATE(Req_I_CC_MaxStale) {
+		parser->cache_control.filled = true;
+		parser->cache_control.dir_allowed = false;
+
 		if (c == '=')
 			__FSM_I_MOVE(Req_I_CC_MaxStaleVBeg);
 		if (IS_WS(c) || c == ',' || IS_CRLF(c)) {
 			req->cache_ctl.max_stale = UINT_MAX;
 			req->cache_ctl.flags |= TFW_HTTP_CC_MAX_STALE;
 			__FSM_I_JMP(Req_I_EoT);
 		}
+		parser->cache_control.dir_flag = TFW_HTTP_CC_OTHER;
+		/* something like "max-staledfgh$!dgh" */
 		__FSM_I_JMP(Req_I_CC_Ext);
 	}
 
 	__FSM_REQUIRE_FIRST_DIGIT(Req_I_CC_MaxStaleVBeg, Req_I_CC_MaxStaleV);
 
 	__FSM_STATE(Req_I_CC_MaxStaleV) {
+		parser->cache_control.filled = true;
+		 /* already done by Req_I_CC_MaxStale though */
+		parser->cache_control.dir_allowed = false;
 		__fsm_sz = __data_remain(p);
 		__fsm_n = parse_uint_list(p, __fsm_sz, &parser->_acc);
 		if (__fsm_n == CSTR_POSTPONE)
 			__msg_hdr_chunk_fixup(data, len);
 		if (__fsm_n < 0)
-			return __fsm_n;
+			__FSM_EXIT(__fsm_n);
 		req->cache_ctl.max_stale = parser->_acc;
 		req->cache_ctl.flags |= TFW_HTTP_CC_MAX_STALE;
 		__FSM_I_MOVE_n(Req_I_EoT, __fsm_n);
 	}
 
 	__FSM_STATE(Req_I_CC_Ext) {
+		/* Any flag we don't understand.
+		 * Here we just skip all the tokens, double quotes and equal signs.
+		 */
 		/* TODO: process cache extensions. */
-		__FSM_I_MATCH_MOVE(qetoken, Req_I_CC_Ext);
-		c = *(p + __fsm_sz);
-		if (IS_WS(c) || c == ',')
-			__FSM_I_MOVE_n(Req_I_EoT, __fsm_sz + 1);
-		if (IS_CRLF(c))
-			return __data_off(p + __fsm_sz);
-		return CSTR_NEQ;
+		__FSM_I_MATCH_MOVE_fixup_finish(qetoken, Req_I_CC_Ext, 0, {
+			parser->cache_control.filled = true;
+			/* Don't check the "dir_allowed" if we are just
+			 * skipping the tail of already recognized flag. */
+			if (parser->cache_control.dir_flag == 0) {
+				if (unlikely(!parser->cache_control.dir_allowed))
+					__FSM_EXIT(TFW_BLOCK);
+				parser->cache_control.dir_flag = TFW_HTTP_CC_OTHER;
+			}
+		});
+		if (__fsm_sz > 0) {
+			if (parser->cache_control.dir_flag == 0) {
+				if (unlikely(!parser->cache_control.dir_allowed))
+					__FSM_EXIT(TFW_BLOCK);
+				parser->cache_control.dir_flag = TFW_HTTP_CC_OTHER;
+			}
+			parser->cache_control.filled = true;
+		}
+		if (parser->cache_control.dir_flag != 0) {
+			/* Req_I_EoT will update the field at comma */
+			parser->cache_control.dir_allowed = false;
+		}
```

Comment:
Hello! It seems I do not understand a conception of `dir_allowed` flag (or can hardly agree with it).

The flag is set to `true` in two states `Req_I_CC_start` and `Resp_I_EoT`, reset to `false` in several other states and is checked in many other states.

It seems that the flag is used to disable (make disallowed) a subset of states.

Is not it better to provide the same by hardcoding transitions between states with `__FSM_JPM`, `__FSM_I_MOVE_xxxx`?

We can control state transition hardly and `__FSM_START` is to restore that state on the next call of the function rather than to enter a random unpredictable state.

Reply:
Looking from more general position, I'd use states to encode parser positions inside text being parsed and flags and variables to store information which was detected in the past and needed in the future. However performance consideration could dictate something different.